### PR TITLE
Remove None at outputs of the model, improve types handling in frontend

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Remove None at outputs of the model*
 - *Add PyNone custom type*
 - *Handle strings and None in FW Node attrs*

### Tickets:
 - *100872*

### CODE REVIEW

**Review by eaidova — APPROVED** [review]

File: src/bindings/python/src/pyopenvino/utils/utils.cpp [hunk 1]
```diff
@@ -242,6 +242,10 @@ ov::Any py_object_to_any(const py::object& py_obj) {
         return py::cast<ov::frontend::type::Tensor>(py_obj);
     } else if (py::isinstance<ov::frontend::type::List>(py_obj)) {
         return py::cast<ov::frontend::type::List>(py_obj);
+    } else if (py::isinstance<ov::frontend::type::Str>(py_obj)) {
+        return py::cast<ov::frontend::type::Str>(py_obj);
+    } else if (py::isinstance<ov::frontend::type::PyNone>(py_obj)) {
+        return py::cast<ov::frontend::type::PyNone>(py_obj);
```
Comment (slyalin) [inline src/bindings/python/src/pyopenvino/utils/utils.cpp:248]: So we continue affecting the common part, but the changes are required for (at this moment) for PT FE only. Any way to decompose functionality of this function and distribute it to more specialized parts like pt fe py bindings?

File: src/frontends/pytorch/src/node_context.cpp [hunk 2]
```diff
@@ -14,6 +14,28 @@ namespace ov {
...
 namespace pytorch {
 
+OutputVector NodeContext::as_constant() const {
+    auto dtype = m_decoder->get_output_type(0);
+    if (dtype.is<type::Str>()) {
+        // Cannot represent string as Constant, creating FrameworkNode
+        auto str = m_decoder->as_string();
+        auto fw_node = std::make_shared<PtFrameworkNode>(m_decoder, OutputVector{});
+        auto attrs = fw_node->get_attrs();
+        attrs["string_value"] = str;
```
Comment (slyalin) [inline src/frontends/pytorch/src/node_context.cpp:24]: Don't understand where we are using it. Is it for future usages when processing of string constants become normal?

File: src/bindings/python/requirements_test.txt [hunk 3]
```diff
@@ -37,4 +37,5 @@ tox
 types-pkg_resources
 wheel>=0.38.1
 protobuf~=3.18.1
-numpy>=1.16.6,<=1.23.4
\ No newline at end of file
+numpy>=1.16.6,<=1.23.4
+torch
```
Comment (rkazants) [inline src/bindings/python/requirements_test.txt:41]: can we set exact version for torch dependency to have deterministic validation?

File: src/frontends/pytorch/src/node_context.cpp (hunk 2 above)
Comment (mvafin) [reply src/frontends/pytorch/src/node_context.cpp:24]: yes, currently we use it nowhere. It can be used to get string value in transformations, but it is also possible to get it directly from decoder.

Comment (mvafin) [reply src/frontends/pytorch/src/node_context.cpp:24]: For example if we can use it to resolve dicts in transformations

File: src/bindings/python/requirements_test.txt (hunk 3 above)
Comment (ilya-lavrenov) [inline src/bindings/python/requirements_test.txt:41]: it's not good to have such dependency in python API tests. 
Python API tests test C++ / Python interaction inside OpenVINO. 

About integration of PyTorch and OpenVINO - can / should it be tested via PyTorch Layer tests? 

@jiwaszki what do you think?

Comment (mvafin) [reply src/bindings/python/requirements_test.txt:41]: can do it layer test, but I see onnx dependency there, and decoder is located there, so it seems like a good place for such tests.

Comment (ilya-lavrenov) [reply src/bindings/python/requirements_test.txt:41]: ONNX will be removed in future, see discussions here https://github.com/openvinotoolkit/openvino/pull/14985

Maybe it's a good time to consider moving Python frontend parts closer to frontends (because when we test frontend, we have to have frameworks, but when we test python API - we definitely don't need frameworks). We have this for C++ where we don't mix code for ngraph, inference and frontends.

Comment (mvafin) [reply src/bindings/python/requirements_test.txt:41]: I agree that makes sense, but I think it should be a different discussion

File: tests/layer_tests/pytorch_tests/test_tuple_construct.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        import torch
+
+        class prim_tuple_construct_single_value(torch.nn.Module):
+
+            def forward(self, x):
+                return (x,)
+
+        class prim_tuple_construct(torch.nn.Module):
+
+            def forward(self, x):
```
Comment (rkazants) [inline tests/layer_tests/pytorch_tests/test_tuple_construct.py:23]: Should we test tuples in parameters to `forward`?

File: tests/layer_tests/pytorch_tests/pytorch_layer_test_class.py [hunk 5]
```diff
@@ -94,10 +94,18 @@ def _test(self, model, ref_net, kind, ie_device, precision, ir_version, infer_ti
             fw_res = (fw_res,)
 
         output_list = list(infer_res.values())
-        assert len(fw_res) == len(
-            output_list), f'number of outputs not equal, {len(fw_res)} != {len(output_list)}'
+
+        flatten_fw_res = []
+        for res_item in fw_res:
+            # if None is at output we skip it
+            if res_item is None:
```
Comment (rkazants) [inline tests/layer_tests/pytorch_tests/pytorch_layer_test_class.py:101]: can you please elaborate in what cases the result can be `None`?

Comment (eaidova) [reply tests/layer_tests/pytorch_tests/pytorch_layer_test_class.py:101]: the most popular case is embedding loss function calculation in model, which used only in training stage. When we move model to eval mode and does not provide annotation, it is not calculated and return by default None.
As example, https://github.com/pytorch/vision/blob/main/torchvision/models/inception.py#L170

Comment (rkazants) [reply tests/layer_tests/pytorch_tests/pytorch_layer_test_class.py:101]: can we add such comment into the code?

File: src/bindings/python/src/pyopenvino/utils/utils.cpp (hunk 1 above)
Comment (mvafin) [reply src/bindings/python/src/pyopenvino/utils/utils.cpp:257]: Not sure how to do it. Those types should be shared between frontends though, so I don't think that this is huge issue

File: tests/layer_tests/pytorch_tests/test_tuple_construct.py (hunk 4 above)
Comment (mvafin) [reply tests/layer_tests/pytorch_tests/test_tuple_construct.py:23]: We test output for now

File: src/bindings/python/requirements_test.txt (hunk 3 above)
Comment (mvafin) [reply src/bindings/python/requirements_test.txt]: Moved to layer tests for now, but I don't like that solution

File: tests/layer_tests/pytorch_tests/pytorch_layer_test_class.py (hunk 5 above)
Comment (mvafin) [reply tests/layer_tests/pytorch_tests/pytorch_layer_test_class.py:101]: Added comment

**Review by rkazants — APPROVED** [review]

File: .ci/azure/linux.yml [hunk 6]
```diff
@@ -516,6 +516,11 @@ jobs:
       $(RUN_PREFIX) python3 -m pytest $(LAYER_TESTS_DIR)/mo_python_api_tests/test_mo_convert_pytorch.py --ir_version=11 --junitxml=./TEST-test_mo_convert_pytorch.xmlTEST
     displayName: 'MO Python API Tests - Import PyTorch model from memory'
 
+  - script: |
+      python3 -m pip install -r $(LAYER_TESTS_DIR)/requirements.txt
+      $(RUN_PREFIX) python3 -m pytest $(LAYER_TESTS_DIR)/py_frontend_tests --junitxml=./TEST-test_mo_convert_pytorch.xmlTEST
```
Comment (ilya-lavrenov) [inline .ci/azure/linux.yml:521]: this step uses the same file name as previous one `TEST-test_mo_convert_pytorch.xmlTEST`. Please, select different name
Secondly, is file extension `xmlTEST` correct?

Comment (mvafin) [reply .ci/azure/linux.yml:521]: I believe extension can be any, but yeah seems weird, will change to xml

Comment (mvafin) [inline .ci/azure/linux.yml:521]:
```suggestion
      $(RUN_PREFIX) python3 -m pytest $(LAYER_TESTS_DIR)/py_frontend_tests --junitxml=./TEST-test_py_fontend.xml
```

**Review by ilya-lavrenov — APPROVED** [review]
